src/compiler/checker.ts

```diff
@@ -20691,7 +20691,7 @@ namespace ts {
             }
 
             function narrowByInKeyword(type: Type, literal: LiteralExpression, assumeTrue: boolean) {
-                if (type.flags & (TypeFlags.Union | TypeFlags.Object) || isThisTypeParameter(type)) {
+                if (type.flags & TypeFlags.Union || type.flags & TypeFlags.Object && declaredType.flags & TypeFlags.Union || isThisTypeParameter(type)) {

```

Comment:
Rather than the declared type, shouldn't this just be looking at the input type? If we look at the declared type, it precludes things like
```ts
declare var x:  unknown;
if (isAOrB(x)) {
  if ("aProp" in x) {
    x; // A
  }
  else if ("bProp" in x) {
    x; // B
  }
  else if ("cProp" in x) {
    x; // never
  }
}
```
since I'm pretty sure we just merged a fix to a lot of cases like that.

Reply:
The fix that just went in was for discriminant property narrowing, which solved examples that are similar to yours, but it's a disjoint code path. Changing `declaredType` to `type` would revert to the original behaviour, but also have the correct behaviour here.

I've made a slight modification that narrows singletons if the current type is not the same as the declared type. This is abit of a hack, but in general working out the intent does not seem easy.

I added some discussion to the linked issue. The best thing might be to make no change at all.